tests/cassette_utils.py

```
@@ -8,13 +8,250 @@
+
+    A new user turn beyond the previous request's history marks a new conversation turn or run (a
+    fresh `agent.run()`), where a legitimately different toolset -- including none -- is expected.
+    Within a single run the agent loop only appends assistant and tool-result messages, never a new
+    user prompt, so a genuine user turn is the reliable boundary signal. Tool and function results are
+    carried on user-role messages by several providers (Anthropic/Bedrock `tool_result`/`toolResult`
+    content, Google `functionResponse` parts, OpenAI Responses `function_call_output`); those are part
+    of the same turn, so they are not counted as a new one.
+    """
+    try:
+        message = json.loads(block)
+    except Exception:
```

Ruff (0.15.21)

[warning] Do not catch blind exception: `Exception`

(BLE001)